File: src/com/nevakanezah/horseenhancer/util/ArgumentParserTest.java
package com.nevakanezah.horseenhancer.util;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

import org.junit.jupiter.api.Test;

class ArgumentParserTest
{
	@Test
	void testBase()
	{
		String[] input = "A B C".split(" ", -1);
		
		Object[] actual = ArgumentParser.parseSplittedArguments(input).toArray();
		
		assertArrayEquals(new String[] {"A", "B", "C"}, actual);
	}
	
	@Test
	void testQuoteEntire()
	{
		String[] input = "\"A B C\"".split(" ", -1);

		Object[] actual = ArgumentParser.parseSplittedArguments(input).toArray();
		
		String[] expected = {"A B C"};

		assertArrayEquals(expected, actual);
	}
	
	@Test
	void testQuotePartial()
	{
		String[] input = "\"A B\" C".split(" ", -1);

		Object[] actual = ArgumentParser.parseSplittedArguments(input).toArray();
		
		String[] expected = {"A B", "C"};

		assertArrayEquals(expected, actual);
	}
	
	@Test
	void testQuoteUnbalancedStart()
	{
		String[] input = "\"A B C".split(" ", -1);

		Object[] actual = ArgumentParser.parseSplittedArguments(input).toArray();
		
		String[] expected = {"\"A", "B", "C"};

		assertArrayEquals(expected, actual);
	}
	
	@Test
	void testQuoteUnbalancedEnd()
	{
		String[] input = "A\" B C".split(" ", -1);

		Object[] actual = ArgumentParser.parseSplittedArguments(input).toArray();
		
		String[] expected = {"A\"", "B", "C"};

		assertArrayEquals(expected, actual);
	}
}

File: src/test/java/com/nevakanezah/horseenhancer/util/ArgumentParserTest.kt
package com.nevakanezah.horseenhancer.util

import org.junit.Test
import kotlin.test.assertContentEquals

internal class ArgumentParserTest {
    @Test
    fun parseSplitArguments() {
        val inputExpectedMap = mapOf(
            "summon help" to listOf("summon", "help"),
            "input1 input2" to listOf("input1", "input2"),
            "\"input1 input2" to listOf("\"input1", "input2"),
            "input1 input2\"" to listOf("input1", "input2\""),
            "\"input1 input2\"" to listOf("input1 input2"),
            "\"input1\" input2" to listOf("input1", "input2"),
            "\"input1 input2\" input3" to listOf("input1 input2", "input3"),
            "\"input1 \"input2 input3" to listOf("\"input1", "\"input2", "input3"),
//            "\"input1 \"input2 input3\"" to listOf("\"input1", "input2 input3"),
        )

        val actual = inputExpectedMap.keys.map { it.split(' ').toTypedArray() }.map(com.nevakanezah.horseenhancer.util.ArgumentParser::parseSplitArguments)

        assertContentEquals(inputExpectedMap.values, actual)
    }
}
